File: C++/AVLTree.cpp
#include "AVLTree.hpp"


AVLTree::AVLTree(int v) {
	this->value = v;
	this->height = 1;
	this->left = this->right = nullptr;
}

void AVLTree::update(AVLTree* tree) {
	this->value = tree->value;
	this->height = tree->height;
	this->left = tree->left;
	this->right = tree->right;
}

unsigned char AVLTree::getHeight(AVLTree* node) {
	return node ? node->height : 0;
}

int AVLTree::bfactor() {
	return getHeight(this->right) - getHeight(this->left);
}

void AVLTree::fixHeight() {
	this->height = std::max(getHeight(this->left), getHeight(this->right)) + 1;
}

void AVLTree::rotateRight() {
	AVLTree* thisCopy = new AVLTree(this->value);
	thisCopy->update(this);

	AVLTree* newTree = thisCopy->left;
	thisCopy->left = newTree->right;
	newTree->right = thisCopy;
	thisCopy->fixHeight();
	newTree->fixHeight();

	this->update(newTree);
}

void AVLTree::rotateLeft() {
	AVLTree* thisCopy = new AVLTree(this->value);
	thisCopy->update(this);
	
	AVLTree* newTree = thisCopy->right;
	thisCopy->right = newTree->left;
	newTree->left = thisCopy;
	thisCopy->fixHeight();
	newTree->fixHeight();
	
	this->update(newTree);
}

void AVLTree::balance() {
	this->fixHeight();
	if (this->bfactor() == 2) {
		if (this->right->bfactor() < 0)
			this->right->rotateRight();
		this->rotateLeft();
	} else if (this->bfactor() == -2) {
		if (this->left->bfactor() > 0)
			this->left->rotateLeft();
		this->rotateRight();
	}
}

void AVLTree::insert(int val) {
	if (val < this->value) {
		if (this->left)
			this->left->insert(val);
		else
			this->left = new AVLTree(val);
	} else if (val > this->value) {
		if (this->right)
			this->right->insert(val);
		else
			this->right = new AVLTree(val);
	}
	this->balance();
}

bool AVLTree::find(int val) {
	if (val < this->value) {
		return this->left ? this->left->find(val) : false;
	} else if (val > this->value) {
		return this->right ? this->right->find(val) : false;
	} else {
		return true;
	}
}

AVLTree* AVLTree::findmin() {
	return this->left ? this->left->findmin() : this;
}

AVLTree* AVLTree::removemin() {
	if (this->left == nullptr)
		return this->right;
	this->left = this->left->removemin();
	this->balance();
	return this;
}

bool AVLTree::remove(int val) {  // нужна чтобы не проверять наличие val при каждом рекурсивном вызове
	if (!this->find(val))
		return false;
	if (this->value == val) {
		// delete this;  // TODO: если оставить delete this, то будет ошибка из-за того что delete странно работает
		return true;
	}
	return this->_remove(val);
}

bool AVLTree::_remove(int val) {
	if (val < this->value) {
		if (this->left->left == nullptr && this->left->right == nullptr) {
			this->left = nullptr;
			return true;
		} else {
			this->left->_remove(val);
		}
	} else if (val > this->value) {
		if (this->right->left == nullptr && this->right->right == nullptr) {
			this->right = nullptr;
			return true;
		} else {
			this->right->_remove(val);
		}
	} else {
		AVLTree* thisCopy = new AVLTree(this->value);
		thisCopy->update(this);
		AVLTree* leftCopy = thisCopy->left;
		AVLTree* rightCopy = thisCopy->right;
		if (rightCopy == nullptr) {
			this->update(leftCopy);
		} else {
			AVLTree* minTree = rightCopy->findmin();
			minTree->right = rightCopy->removemin();
			minTree->left = leftCopy;
			this->update(minTree);
		}
	}
	this->balance();
	return true;
}

std::string AVLTree::getString() {
	std::string leftInorder = this->left != nullptr ? this->left->getString() : "";
	std::string rightInorder = this->right != nullptr ? this->right->getString() : "";
	if (leftInorder.length() > 0) {
		leftInorder = "(" + leftInorder + ") ";
	}
	if (rightInorder.length() > 0) {
		rightInorder = " (" + rightInorder + ")";
	}
	return leftInorder + std::to_string(this->value) + rightInorder;
}

AVLTree::~AVLTree() {
	delete this->left;
	delete this->right;
}

File: AvlPython/AVLtree.py
class AVLTree:
    def __init__(self, val):
        self.val = val
        self.height = 1
        self.left = self.right = None

    def update(self, other):
        self.val = other.val
        self.height = other.height
        self.left = other.left
        self.right = other.right

    @staticmethod
    def getHeight(node):
        return node.height if node else 0

    def _bfactor(self):
        return self.getHeight(self.right) - self.getHeight(self.left)

    def _fixHeight(self):
        self.height = max(self.getHeight(self.left), self.getHeight(self.right)) + 1

    def _rRight(self):
        tree = AVLTree(self.val)
        tree.update(self)

        newTree = tree.left
        tree.left = newTree.right
        newTree.right = tree
        tree._fixHeight()
        newTree._fixHeight()

        self.update(newTree)

    def _rLeft(self):
        tree = AVLTree(self.val)
        tree.update(self)

        newTree = tree.right
        tree.right = newTree.left
        newTree.left = tree
        tree._fixHeight()
        newTree._fixHeight()

        self.update(newTree)

    def _balance(self):
        self._fixHeight()
        if (self._bfactor() == 2):
            if (self.right._bfactor() < 0):
                self.right._rRight()
            self._rLeft()
        elif (self._bfactor() == -2):
            if (self.left._bfactor() > 0):
                self.left._rLeft()
            self._rRight()

    def findmin(self):
        return self.left.findmin() if self.left else self

    def removemin(self):
        if self.left == None:
            return self.right
        self.left = self.left.removemin()
        self._balance()
        return self

    def insert(self, val):
        if val < self.val:
            if self.left:
                self.left.insert(val)
            else:
                self.left = AVLTree(val)
        elif val > self.val:
            if self.right:
                self.right.insert(val)
            else:
                self.right = AVLTree(val)
        self._balance()

    def find(self, val):
        if val < self.val:
            return self.left.find(val) if self.left else False
        elif val > self.val:
            return self.right.find(val) if self.right else False
        else:
            return True

    def remove(self, val):
        if not self.find(val):
            return False
        if self.height == 1:
            return True
        else:
            return self._remove(val)

    def _remove(self, val):
        if val < self.val:
            if self.left.left == None and self.left.right == None:
                self.left = None
                return True
            else:
                self.left._remove(val)
        elif val > self.val:
            if self.right.left == None and self.right.right == None:
                self.right = None
                return True
            else:
                self.right._remove(val)
        else:
            tree = AVLTree(self.val)
            tree.update(self)
            l = tree.left
            r = tree.right
            if (r == None):
                self.update(l)
            else:
                m = r.findmin()
                m.right = r.removemin()
                m.left = l
                self.update(m)
        self._balance()
        return True

    def __str__(self):
        leftIn = str(self.left) if self.left else ""
        rightIn = str(self.right) if self.right else ""
        if len(leftIn) > 0:
            leftIn = "(" + leftIn + ")"
        if len(rightIn) > 0:
            rightIn = "(" + rightIn + ")"
        return leftIn + str(self.val) + rightIn

    def __del__(self):
        del self


def main():
    tree = AVLTree(1)
    tree.insert(2)
    tree.insert(5)
    tree.insert(6)
    tree.insert(3)
    print(tree)
    tree.remove(5)
    print(tree)
    tree.remove(2)
    print(tree)
    tree.remove(3)
    print(tree)
    tree.remove(6)
    print(tree)
    tree.remove(1)
    print(tree)
    for i in [-30714, 39586, 33217, 34756, 22403, 24117, 34261, -49443, -23765, 23920, 19227, -29927, -35293, 30612]:
        tree.insert(i)
    print(tree)
    print(tree.find(2))
    print(tree.remove(2))
    print(tree)


if __name__ == '__main__':
    main()
